can-bus/common.c

```diff
@@ -0,0 +1,217 @@
+}
+
+/* Our CAN bus receive function */
+int can_receive(uint8_t data[8], int *length) {
+    int nbytes;
+    struct can_frame frame;
+    struct pollfd p[1];
+
+	p[0].fd = sock;
+	p[0].events = POLLIN;
+
+    /* Poll for new data */
+	int retval = poll(p, 1, 10);
+
+    if (retval < 0) {
+        perror("Poll error\n");
+        return 1;
+    }
+    else if (retval == 0) {
+        // No data
+        *length = 0;
+        return EAGAIN;
+    }
+
+    /* Read in the frame data */
+    nbytes = read(sock, &frame, sizeof(struct can_frame));
+    if (nbytes < 0) {
+       perror("Read error\n");
+       return 1;
+    }
+    memcpy(data, frame.data, frame.can_dlc);
+    *length = frame.can_dlc;
+    return 0;
+}
+
+/* Connect to the CAN bus */
+int can_connect(const char *address, uint16_t filter)
+{
+    struct can_filter rfilter[1];
+    rfilter[0].can_id   = filter;
+    rfilter[0].can_mask = 0xFFF;
+
+    if ((sock = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
+       perror("Socket open error\n");
+       return 1;
+    }
+    setsockopt(sock, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter));
+    struct ifreq ifr;
+
+    strcpy(ifr.ifr_name, address);
+    ioctl(sock, SIOCGIFINDEX, &ifr);
+
+    struct sockaddr_can addr;
+    memset(&addr, 0, sizeof(addr));
+    addr.can_family = AF_CAN;
+    addr.can_ifindex = ifr.ifr_ifindex;
+    if (bind(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+       perror("Bind error\n");
+       return 1;
+    }
+
+    /* Setup the receive copy buffer */
+    copy_buf = malloc(REC_BUF_SIZE);
+    copy_buf_ptr = copy_buf;
+    copy_buf_len = 0;
+    if (!copy_buf) {
+        printf("Copy buf malloc fail\n");
+        return -99;
+    }
+
+    return 0;
+}
+
+void can_close()
+{
+    free(copy_buf);
+    close(sock);
+}
+
+/* The wolfSSL send packet callback. This function sends the data via ISO-TP
+ * and then listens and processes the ISO-TP control packets until the message
+ * is completely sent */
+int send_ssl(WOLFSSL *ssl, char *buf, int sz, void *ctx)
+{
+    uint8_t data[8];
+    int length;
+    IsoTpLink *g_link = (struct IsoTpLink*)ctx;
+    int ret = isotp_send(g_link, buf, sz);
+    printf("Sending %d bytes\n", sz);
+
+    if (ret) {
+        printf("Error from isotp: %d\n", ret);
+        return -64;
+    }
+
+    while(g_link->send_status != ISOTP_SEND_STATUS_IDLE) {
+        ret = can_receive(data, &length);
+        if (!ret && (length > 0)) {
+            // Got a control frame
+            isotp_on_can_message(g_link, data, length);
+        } else if (ret != EAGAIN) {
+            perror("Receive error\n");
+        }
+        isotp_poll(g_link);
+    }
+
+    return sz;
+}
+
+/* Receive callback for wolfSSL. ISO-TP will only return a buffer once and it
+ * can be more data than wolfSSL wants, so we copy this into an intermediate
+ * buffer and just return what wolfSSL is asking for */
+int recv_ssl(WOLFSSL* ssl, char* buf, int sz, void* ctx)
+{
+    uint8_t data[8];
```

Comment:
Can you make the magic 8 data size a macro or enum? I suspect this is the max CAN payload?